### ISSUE TITLE
Allow users to restart a service

### ISSUE THREAD

**Issue by MarcSkovMadsen** [jupyterhub/jupyter-server-proxy#264]
I am and will be enabling users to run Bokeh, Dash, Flask, Panel, Streamlit and Voila apps on Jupyter Hub and Binder.

Those apps are a part of the explorative environment and users would expect to be able to update the underlying code and experience the updated apps. In my case the users might not even update the code them selves but the app code would be updated by other users/ a system team in one or more shared folders that are mounted.

Currently I don't see any way of restarting those services. I would like a feature to easily restart a service.

## Additional Context

Examples of servers are

- [jupyter-panel-proxy](https://github.com/holoviz/jupyter-panel-proxy)
- [custom Panel HighCharts](https://github.com/MarcSkovMadsen/panel-highcharts/blob/main/binder/jupyter-panel-highcharts-server/jupyter_panel_highcharts_server.py)

I have a similar request for `jupyter-panel-proxy` that it should provide functionality to restart. But I also created a feature request here as I believe the use case and request is more general. See https://github.com/holoviz/jupyter-panel-proxy/issues/14

**Comment by manics** [comment]
Would this PR cover your needs? https://github.com/jupyterhub/jupyter-server-proxy/pull/215

**Comment by bollwyvl** [comment]
I think dedicated restarting/status/logging is worth pursuing. Over [on jupyter(lab)-lsp](https://github.com/jupyter-lsp/jupyterlab-lsp/issues/681), we've had similar concerns, especially for language servers that have file-based configurations someone may wish to update interactively.... or that get wedged, and need hard restarts.

While we previously had a dependency on `jupyter-server-proxy`, we ended up needing more bells and whistles, and have been handling our own stuff for a while. We have a [stoppable server session](https://github.com/jupyter-lsp/jupyterlab-lsp/blob/master/python_packages/jupyter_lsp/jupyter_lsp/session.py#L95) concept... but haven't actually ended up exposing it to users. 

Perhaps we could consider some way of bringing some parts of the codebase back together, as a common upstream dependency, for example:

- in python
  - service status/stop handlers
  - logging handlers
- in (retro)lab
  - a streaming/filtered log viewer 
    - text in xtermjs
    - structured logs in datagrid
  - "Running" UIs (with stop, etc)

ideally, these _wouldn't_ specify how the processes actually work: we're using a mishmash of stuff (especially on windows), and might end up having to do some more with e.g. `anyio`

**Comment by manics** [comment]
@bollwyvl Do you reckon we should close https://github.com/jupyterhub/jupyter-server-proxy/pull/215 and start afresh with your above suggestion?

**Comment by bollwyvl** [comment]
Nah, a couple restart policies (especially if we don't have to think them
up) still seem quite useful... But sometimes it will make sense to want to
drive the bus manually, especially (again) if we don't have to think up a
novel ui.
